### ISSUE TITLE
Adding Dynamic erf (Derf) pointwise function to norms, successor to Dynamic tanh

### ISSUE THREAD

**Issue by coder0143** [google/flax#5457]
### Summary

A recent paper from Princeton, NYU and CMU shows improvements to Dynamic Tanh paper from META (1 common author) showing better performance than Layernorm, RMSNorm and DyT on various tasks. The PR from this issue aims to implement the function into nnx src (norms) with minimal code and appropriate testing. 

A successor to given issue: #5456 

Related links:
Code: https://github.com/zlab-princeton/Derf
Paper: https://arxiv.org/pdf/2512.10938
Title: Stronger Normalization-Free Transformers

cc: @cgarciae @vfdev-5 @samanklesaria 

### Additional context

<img width="1205" height="868" alt="Image" src="https://github.com/user-attachments/assets/83e45e29-01a7-411d-a192-53bd7f24da38" />

**Comment by vfdev-5** [comment]
@coder0143 thanks for suggestion. Do they provide a table to compare LN vs DyT vs Derf in term of flops?
Another question is about follow-up papers/project where transformer architecture is adopting these norm-free idea and shipping trained models for further usage. I would like to see if this technique is being adopted (keeping in mind that it is rather recent paper) and not just used once in the original paper and will be almost abandoned in the next years.

**Comment by Ashutosh0x** [comment]
Hi @vfdev-5, great questions! I'd like to help implement this alongside #5456. Let me address your points:

**1. FLOPs Comparison (LN vs DyT vs Derf):**

Both DyT and Derf are strictly cheaper than LayerNorm because they eliminate the mean/variance reduction operations entirely. From the papers:

| Operation | Per-element Cost | Reduction Required |
|---|---|---|
| LayerNorm | mean + var + normalize + affine | Yes (across features) |
| DyT | tanh + scale + shift | No |
| Derf | erf + scale + shift | No |

DyT and Derf have nearly identical FLOPs — the only difference is 	anh vs erf, both of which are single-instruction on modern accelerators (TPU MXU, GPU SFU). The Derf paper (Table 1, Section 4.1) shows Derf matches DyT throughput while achieving better accuracy on ViT-B/L and GPT-2 benchmarks.

**2. Adoption beyond the original papers:**

- **DyT** (Meta FAIR, March 2025): The official repo has 2.7k+ GitHub stars. It's been integrated into community projects and has active discussion in the JAX/PyTorch ecosystem.
- **Derf** (Princeton/NYU/CMU, Dec 2025): Newer, but builds directly on DyT's foundation with the same research group (1 shared author). The key insight — replacing statistical normalization with learned pointwise functions — is well-validated across architectures (ViT, LLMs, diffusion models, speech).

Both techniques are minimal additions (< 30 lines each) with clear value as building blocks for the NNX module library. I'll submit a PR covering both shortly.

**Comment by coder0143** [comment]
@vfdev-5 Thanks for checking this out. 

Answering your queries - 
1. Yup a flops comparison is given in the paper, the compute cost is equivalent between all three - LN, DyT, Derf. They have used custom trition kernels. The 

<img width="1638" height="843" alt="Image" src="https://github.com/user-attachments/assets/661374d3-9012-4bad-b7c3-6fcd7b82634d" /><br/>

If you would like, I have recently recieved TPU Research Cloud fellowship from Google Research, and can setup a test for custom pallas kernels for all of them. 

2. These papers have actually built upon newer ideas of normalization free transformers, which can help models via:
- Reduction operations
- Variance computation
- Synchronization-heavy ops
This matters enormously for:
- TPUs
- sequence parallelism
- FlashAttention pipelines
- distributed inference

Some newer papers discussing this: 
 - When Does Removing LayerNorm Help? Activation Bounding as a Regime-Dependent Implicit Regularizer (https://arxiv.org/abs/2604.23434)
- On the Mathematical Relationship Between Layer Normalization and Dynamic Activation Functions (https://arxiv.org/abs/2503.21708)

I have been an active contributor to google's [jax-ml/bonsai](https://github.com/jax-ml/bonsai/tree/main) and would like to make a PR to add both the norms through same or different PRs if you would like, Thanks!

@Ashutosh0x I would like to work on this PR, if you don't mind, thanks!

**Comment by vfdev-5** [comment]
@coder0143 yes, I also just seen the Appendix F where the perfs of triton implementation can be comparable.
Can you make benchmarks on GPU using flax LN and DyT, Derf implemented in Jax?

In term of adoption, I think the most prominent paper from the list of citations using Derf is https://arxiv.org/pdf/2604.21999
but still hard to judge how relevent it could be...

Finally, it is not a competition who will implement this and how quickly. The important point to keep in mind is that the new code maintenance has its cost, so it is not the goal to implement every new idea, but new prominent idea.

**Comment by Ashutosh0x** [comment]
@coder0143 Absolutely, go for it! I'll focus on #5456 (DyT) and you take #5457 (Derf) — that way we both ship clean, focused PRs and @vfdev-5 gets two independent reviews rather than one monolithic one.

Happy to cross-reference your implementation for API consistency once we're both in review. Good luck with the benchmarks! 🙌

**Comment by coder0143** [comment]
@vfdev-5 Thanks for your thoughtful feedback. One more paper I have found here is: https://arxiv.org/pdf/2604.01563, 
"Does Your Optimizer Care How You Normalize? Normalization-Optimizer Coupling in LLM Training". 
- Here there is an in depth comparison of RMSNorm, DyT and Derf with various optimizers and results. (Upto 1B) 
- I will benchmark both the implementations on mine (rtx 3050 4gb, archlinux) and cloud (L4 24gb, ubuntu) gpus.

@Ashutosh0x Thanks!!

**Comment by vfdev-5** [comment]
We discussed internally and we may not want to implement Derf or DyT for the reason that we want to keep the library small and Derf/DyT are not yet universal things like Conv, BN, MHA, etc.

For those who would like to use Derf in their work with jax/flax they can use the following implementation:
```python
class DerfNorm(nnx.Module):
    def __init__(self, features: int, rngs: nnx.Rngs):
        self.alpha = nnx.Param(jax.random.normal(rngs(), (features,)) * 0.02 + 1.0)
        self.s = nnx.Param(jnp.zeros((features,)))

    def __call__(self, x: jax.Array) -> jax.Array:
        return jax.scipy.special.erf(self.alpha * x + self.s)
```
Source: https://github.com/che-shr-cat/utm-jax/blob/ae4aa0eda49fa01ce21d7b8e6ba2f912c280db3d/models/layers.py#L16-L22

Thanks everyone for the discussion!
I'll close the issue as not planned.
